### PR TITLE
[FLINK-36455] Sinks retry synchronously

### PR DESCRIPTION
**Pull request by AHeise**
## What is the purpose of the change

Sinks so far retried asynchronously to increase commit throughput in case of temporary issues. However, the contract of notifyCheckpointCompleted states that checkpoints must be side-effect free meaning all transactions have to be committed on return of the PRC call.

## Brief change log

* This commit retries a fixed number of times and then fails in notifyCheckpointCompleted.
* Simplifies parts of committable handling now that all committables of a subtask either succeed or fail

## Verifying this change

* Already covered by many tests
* Adjusted and changed tests in
api/connector/sink2
runtime/operators/sink/committables

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (**yes** / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/streaming/api/connector/sink2/CommittableSummary.java [hunk 1]
```diff
@@ -39,11 +39,21 @@ public class CommittableSummary<CommT> implements CommittableMessage<CommT> {
...
     /** The number of committables coming from the given subtask in the particular checkpoint. */
     private final int numberOfCommittables;
+
+    @Deprecated
     /** The number of committables that have not been successfully committed. */
     private final int numberOfPendingCommittables;
+
+    @Deprecated
     /** The number of committables that are not retried and have been failed. */
     private final int numberOfFailedCommittables;
```
Comment (fapaul) [inline flink-runtime/src/main/java/org/apache/flink/streaming/api/connector/sink2/CommittableSummary.java:49]: Why do we deprecate failed committables? I still see some value to provide the possibility to discard unrecoverable committables. The change also looks unrelated to the retry mechanism.

File: flink-runtime/src/main/java/org/apache/flink/streaming/api/connector/sink2/GlobalCommitterOperator.java [hunk 2]
```diff
@@ -95,6 +94,7 @@
 public class GlobalCommitterOperator<CommT, GlobalCommT> extends AbstractStreamOperator<Void>
         implements OneInputStreamOperator<CommittableMessage<CommT>, Void> {
 
+    private static final int MAX_RETRIES = 10;
```
Comment (fapaul) [inline flink-runtime/src/main/java/org/apache/flink/streaming/api/connector/sink2/GlobalCommitterOperator.java:97]: IMO we should make this a Flink config specific to the sink and allow users to opt-out with the failed committable mechanism. 

Currently the number is duplicated in committer and global committer.

File: flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterOperator.java [hunk 3]
```diff
@@ -164,41 +165,37 @@ public void notifyCheckpointComplete(long checkpointId) throws Exception {
...
         }
     }
 
-    private void retryWithDelay() {
-        processingTimeService.registerTimer(
-                processingTimeService.getCurrentProcessingTime() + RETRY_DELAY,
-                ts -> commitAndEmitCheckpoints());
+    private void emit(CheckpointCommittableManager<CommT> committableManager) {
+        int subtaskId = getRuntimeContext().getTaskInfo().getIndexOfThisSubtask();
+        int numberOfSubtasks = getRuntimeContext().getTaskInfo().getNumberOfParallelSubtasks();
```
Comment (fapaul) [inline flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterOperator.java:187]: Nit: I know this wasn't introduced by this PR but why do we need to fetch the subtask id and number of tasks on every `emit` call?

**Review by fapaul — COMMENTED** [review]
Thanks for making this change. I left a few comments, but I have two main concerns.

- I wonder whether we should leave the failed comittable tracker
- The PR misses changes to the committer interface e.g. `CommitRequest#signalFailedWithKnownReason`

File: flink-runtime/src/main/java/org/apache/flink/streaming/api/connector/sink2/CommittableSummary.java (hunk 1 above)
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/streaming/api/connector/sink2/CommittableSummary.java:49]: Good point. I probably should motivate that change in the commit message first. The short answer is: it doesn't work. Let's go to the long one.

So far, we increase this number on known issue and don't emit the respective committable. That means that the global committer would need to wait for all committables except the failed once and then commit. However, it always used to wait for all known committables running in an infinite loop.

The change of this PR only creates the summary of successful commits. Unknown errors cause a restart loop and known errors cause the committables to be dropped from the statistics. So the global committer waits for all committables of the summary and works now.

The alternative would be to still update the stastics as you propose and ignore the failed in the global committer. However, I wonder what the value of that is. If you'd like to keep it just to have less disruptions, I can revert the change and fix it.

However, the custom topologies that I have seen so far, also run into the same issue as the global committer (and now I think that the compactor has the same issues). I think that emitting the stats on the ignored committables ultimately just increases complexity without giving downstream operators any good handle whatsoever.

WDYT?

File: flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterOperator.java (hunk 3 above)
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterOperator.java:189]: What's the concern here? Peformance? JVM should inline the call to pretty much result into a field access.

**Comment by AHeise** [comment]
Reverted the deprecation of numFailed and added a config option for the retries. PTAL @fapaul

File: flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterOperator.java (hunk 3 above)
Comment (fapaul) [reply flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterOperator.java:189]: Mostly should looked strange.

File: flink-core/src/main/java/org/apache/flink/configuration/SinkOptions.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    /**
+     * The number of retries on a committable (e.g., transaction) before Flink application fails and
+     * potentially restarts.
+     */
+    public static final ConfigOption<Integer> COMMITTER_RETRIES =
+            key("sink.committer.retries")
+                    .intType()
+                    .defaultValue(10)
+                    .withDescription(
+                            "The number of retries on a committable (e.g., transaction) before Flink application fails and potentially restarts.");
```
Comment (fapaul) [inline flink-core/src/main/java/org/apache/flink/configuration/SinkOptions.java:37]: Nit: Should we mention here that it applies to `CommitRequest#signalFailedWithUnknownReason` and not generic exceptions that are thrown during calling `commit` on the committer.

File: flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CheckpointCommittableManager.java [hunk 5]
```diff
@@ -69,8 +61,12 @@ public interface CheckpointCommittableManager<CommT> {
...
-     * @return successfully committed committables with meta information
+     * @param maxRetries
      */
-    Collection<CommittableWithLineage<CommT>> commit(Committer<CommT> committer)
+    void commit(Committer<CommT> committer, int maxRetries)
             throws IOException, InterruptedException;
+
+    Collection<CommT> getSuccessfulCommittables();
+
+    int getNumFailed();
```
Comment (fapaul) [inline flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CheckpointCommittableManager.java:71]: Please add doc strings to the method

**Review by fapaul — APPROVED** [review]
LGTM now, can you also add a small note to the release notes that we introduce a new config

File: flink-core/src/main/java/org/apache/flink/configuration/SinkOptions.java (hunk 4 above)
Comment (AHeise) [reply flink-core/src/main/java/org/apache/flink/configuration/SinkOptions.java:37]: A good point. I fear that giving too much details is confusing (which end user knows CommitRequest?).

I'd rephrase to 
`The number of retries a Flink application attempts for committable operations (such as transactions) on retriable errors, as specified by the sink connector, before Flink fails and potentially restarts.`

Comment (fapaul) [reply flink-core/src/main/java/org/apache/flink/configuration/SinkOptions.java:37]: Sounds much better 👍
